### PR TITLE
CI: Adding Appveyor testing

### PR DESCRIPTION
**Pull request by thopiekar**
This adds a recipe for Appveyor how to make our tests. This one is only making tests using Python3 and they prove that currently SDK lookup is broken when using Visual Studio 2017.
Therefore I'm also working now on adding a proper lookup for it, too.
(This is just the beginning, but can be merged already, of course.)

### CODE REVIEW

**Comment by thopiekar** [comment]
Now together with #1160 and https://ci.appveyor.com/project/thopiekar/pywin32/build/1.0.11/job/yvgpjpn51axdsda9 it is proven that this PR works.

**Comment by mhammond** [comment]
> Do you want to enable appveyor for this repository?

I don't have any objections to that, but I don't know much about it. I'm a little confused by vs2017 though - it appears 2015 is the latest that any Python version is actually built with?

**Comment by xoviat** [comment]
@mhammond Do you know the steps to sign up for an appveyor account?

**Comment by xoviat** [comment]
Just go to https://ci.appveyor.com/login, and use the github login. Then add this repository.

**Comment by thopiekar** [comment]
Hmm, I can switch to VS2015 and retry. I'm not that much experienced in the different MS programming environments.

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  # This step builds your wheels.
+  # Again, you only need build.cmd if you're building C extensions for
+  # 64-bit Python 3.3/3.4. And you need to use %PYTHON% to get the correct
+  # interpreter
+  - "CALL \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" %WIN_ARCH%"
+  - "%PYTHON%\\python.exe %SETUP_PY% bdist_wheel"
+
+artifacts:
+  # bdist_wheel puts your built wheel in the dist directory
+  - path: dist\*
```
Comment (webknjaz) [inline appveyor.yml]: You could also do wheels deployment to PYPI via Twine

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  # to put the Python version you want to use on PATH.
+  - "%PYTHON%\\python.exe pywin32_testall.py"
+
+after_test:
+  # This step builds your wheels.
+  # Again, you only need build.cmd if you're building C extensions for
+  # 64-bit Python 3.3/3.4. And you need to use %PYTHON% to get the correct
+  # interpreter
+  - "CALL \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" %WIN_ARCH%"
+  - "%PYTHON%\\python.exe %SETUP_PY% bdist_wheel"
```
Comment (webknjaz) [inline appveyor.yml]: Appveyor is capable of showing you a list of tests separately from the console output:
```yaml
after_test:
  - ps: |
      $wc = New-Object 'System.Net.WebClient'
      $wc.UploadFile("https://ci.appveyor.com/api/testresults/junit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path .\junit-test-results.xml))  # make sure your test runner puts junit-compatible test results into this file
```

File: appveyor.yml (hunk 1 above)
Comment (webknjaz) [reply appveyor.yml]: This is how:
```yaml
deploy_script:
  - ps: >-
      if($env:appveyor_repo_tag -eq 'True') {
          Invoke-Expression "python -m twine upload dist/* --username $env:PYPI_USERNAME --password $env:PYPI_PASSWORD"
      }
```

File: appveyor.yml (hunk 2 above)
Comment (thopiekar) [reply appveyor.yml]: Sounds interesting! Personally I don't have much experiences with Appveyor, but do you know how it comes that I'm getting `MemoryError`s? In my point of view I'm doing the same steps, like in my local setup, but at Appveyor I'm getting these errors.

Comment (webknjaz) [reply appveyor.yml]: Dunno, you probably hit some memory limit there. I don't have insights on this, but you should be able to report this to Appveyor crew.

Comment (webknjaz) [reply appveyor.yml]: @thopiekar BTW, I think you don't need `setup3.py`: it's been possible to write cross-compatible code for ages. Just fix a few bits of syntax (print etc?) and imports (try/except ImportError + drop distutils in favour of modern setuptools) in `setup.py`

Comment (thopiekar) [reply appveyor.yml]: I know. My intention was to get this CI functionality into pywin32 first and then begin to contribute with cleanups and updating some PRs I already made.
I think it is also very useful for other people to know early whether their changes work on all supported python versions or not.

Comment (webknjaz) [reply appveyor.yml]: you can still enable Appveyor tests, but don't make failing CI prevent PR from merges until this is fixed

Comment (webknjaz) [reply appveyor.yml]: https://www.appveyor.com/docs/build-environment/#build-vm-configurations it seems they provide you with enough memory

Comment (thopiekar) [reply appveyor.yml]: Strange, so either a memory leak on PyWin32's side or something else I missed or don't see on my own.

Comment (webknjaz) [reply appveyor.yml]: memory leak looks happening in `find_platform_sdk_dir` function, maybe it's related to the fact you do `exec`, i don't know. I'd start with just logging memory usage at various points + removing `exec()` in favor of native script execution. btw, I'd disable creation of `*.pyc` files for test purposes as well. (via env var)

**Comment by thopiekar** [comment]
Thank you for looking into it. Since you mention it, it could be a look, which is running all the time and does not want to stop. If I would know where the MSSDK is located on Appveyor, I would set it, but currently I'm counting on the setup.py script to do this for me.

**Comment by thopiekar** [comment]
Updates from my side:
* I'm experimenting now in a different branch and will upload all work as soon as I have something fully working.
* >= 3.5 builds are working, but without tests, because we have py2-only scripts it seems.

Will continue experimenting and try to get everything built.

**Comment by thopiekar** [comment]
Managed to get a bit further again:
* Builds for Python 3.5 and 3.6 work fine.
* Some builds can't find the MFC libraries. I noticed that the lookup paths are hardcoded in our setup.py. Will replace them with setuptools functions.
* Finally, I have a more worrying issue when building for Python < 3.5. I get syntax errors, which are pointing to headers within the MS SDK. Contacted Appveyor and someone is looking onto it. Do you know how it comes?
https://ci.appveyor.com/project/thopiekar/pywin32/build/1.0.47/job/sigjb8p0swqx5uqq
@webknjaz maybe?

**Comment by webknjaz** [comment]
IDK, I'm not a windows guy

**Comment by thopiekar** [comment]
Ah, shame. How about @mhammond ? Any ideas?

Hope it is only an issue on Appveyor's side and I can continue with patches around the MFC lookup functions.

**Comment by thopiekar** [comment]
Continued at: #1187

**Pull request closed without merging**